src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java

```diff
@@ -0,0 +1,282 @@
+* THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+package htsjdk.variant.vcf;
+
+import java.io.BufferedInputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.channels.SeekableByteChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.function.Function;
+import java.util.zip.GZIPInputStream;
+
+import htsjdk.samtools.seekablestream.SeekablePathStream;
+import htsjdk.samtools.util.AbstractIterator;
+import htsjdk.samtools.util.CloseableIterator;
+import htsjdk.samtools.util.CloserUtil;
+import htsjdk.samtools.util.IOUtil;
+import htsjdk.tribble.readers.AsciiLineReader;
+import htsjdk.tribble.readers.AsciiLineReaderIterator;
+import htsjdk.tribble.readers.PositionalBufferedStream;
+import htsjdk.tribble.util.ParsingUtils;
+import htsjdk.variant.bcf2.BCF2Codec;
+import htsjdk.variant.bcf2.BCFVersion;
+import htsjdk.variant.variantcontext.VariantContext;
+
+/**
+ * A Class building {@link htsjdk.variant.vcf.VCFIterator}
+ * 
+ * Example:
+ * 
+ * <pre>
+ * VCFIterator r = new VCFIteratorBuilder().open(System.in);
+ * while (r.hasNext()) {
+ *     System.out.println(r.next());
+ * }
+ * r.close();
+ * </pre>
+ * 
+ * @author Pierre Lindenbaum / @yokofakun
+ * @see {@link htsjdk.variant.vcf.VCFIterator}
+ *
+ */
+
+public class VCFIteratorBuilder {
+
+    /**
+     * creates a VCF iterator from an input stream It detects if the stream is a
+     * BCF stream or a GZipped stream.
+     * 
+     * @param in inputstream
+     * @return the VCFIterator
+     * @throws IOException
+     */
+    @SuppressWarnings("static-method")
+    public VCFIterator open(final InputStream in) throws IOException {
+        if (in == null) {
+            throw new IllegalArgumentException("input stream is null");
+            }
+        // sizeof a BCF header (+ min/max version)
+        final int sizeof_bcf_header =  BCFVersion.MAGIC_HEADER_START.length + 2*Byte.BYTES;
+        // wrap the input stream into a BufferedInputStream to reset/read a BCFHeader or a GZIP
+        // buffer must be large enough to contain the BCF header and/or GZIP signature
+        BufferedInputStream  bufferedinput = new BufferedInputStream(in, Math.max(
+               sizeof_bcf_header,
+               IOUtil.GZIP_HEADER_READ_LENGTH
+               ));
+        // test for gzipped inputstream 
+        if(IOUtil.isGZIPInputStream(bufferedinput)) {
+            // this is a gzipped input stream, wrap it into GZIPInputStream
+            // and re-wrap it into BufferedInputStream so we can test for the BCF header
+            bufferedinput = new BufferedInputStream(
+                    new GZIPInputStream(bufferedinput),
+                    sizeof_bcf_header
+                    );
+        }
+
+        // try to read a BCF header
+        bufferedinput.mark(sizeof_bcf_header);
+        final BCFVersion bcfVersion = BCFVersion.readBCFVersion(bufferedinput);
+        bufferedinput.reset();
```

Comment:
I'd lobby to move this BCF-sniffing logic out into a public method somewhere, probably in BCFCodec; maybe to a static method that returns the BCFCodec or null on failure. In addition to keeping the knowledge of the BCF internals inside the BCFCodec, it would be immediately useful to solve [this problem](https://github.com/HadoopGenomics/Hadoop-BAM/issues/132) in Hadoop-BAM.